### PR TITLE
[Dependency Scanning] Collect and report each module dependency's Link Libraries

### PR DESCRIPTION
**Pull request by artemcm**

### CODE REVIEW

**Comment by artemcm** [comment]
https://github.com/apple/llvm-project/pull/8810
@swift-ci test

File: include/swift/AST/ModuleDependencies.h [hunk 1]
```diff
@@ -634,22 +638,45 @@ class ModuleDependencyInfo {
...
+    storage->linkLibraries.assign(linkLibraries.begin(), linkLibraries.end());
+  }
+
+  bool isStaticLibrary() const {
+    if (auto *detail = getAsSwiftInterfaceModule())
+      return detail->isStatic;
+    else if (auto *detail = getAsSwiftBinaryModule())
+      return detail->isStatic;
+    // TODO: Can Clang module deps indicate they correspond
+    // to a static library?
```
Comment (ian-twilightcoder) [inline include/swift/AST/ModuleDependencies.h:676]: No, module maps outside of a framework are only loosely associated with a library via a `link` statement, but that just has the library name and not its type or extension.

File: lib/AST/ConformanceLookupTable.cpp [hunk 2]
```diff
@@ -306,9 +306,9 @@ void ConformanceLookupTable::updateLookupTable(NominalTypeDecl *nominal,
             if (!proto)
               continue;
             auto kp = proto->getKnownProtocolKind();
-            assert(!found.isSuppressed ||
-                   kp.has_value() &&
-                       "suppressed conformance for non-known protocol!?");
+//            assert(!found.isSuppressed ||
+//                   kp.has_value() &&
+//                       "suppressed conformance for non-known protocol!?");
```
Comment (ian-twilightcoder) [inline lib/AST/ConformanceLookupTable.cpp:311]: Was this an intentional change?

File: lib/AST/ModuleDependencies.cpp [hunk 3]
```diff
@@ -184,6 +184,7 @@ void ModuleDependencyInfo::addModuleImports(
   sourceFile.getTopLevelDecls(decls);
   for (auto decl : decls) {
     auto importDecl = dyn_cast<ImportDecl>(decl);
+    //auto unboundImport = UnboundImport(importDecl);
```
Comment (ian-twilightcoder) [inline lib/AST/ModuleDependencies.cpp:187]: Don't need this?

File: lib/AST/ModuleDependencies.cpp [hunk 4]
```diff
@@ -509,6 +510,86 @@ swift::dependencies::checkImportNotTautological(const ImportPath::Module moduleP
...
 
+void
+swift::dependencies::registerCxxInteropLibraries(
+    const llvm::Triple &Target,
+    StringRef mainModuleName,
+    bool hasStaticCxx, bool hasStaticCxxStdlib,
+    std::function<void(const LinkLibrary&)> RegistrationCallback) {
+  if (Target.isOSDarwin())
+    RegistrationCallback(LinkLibrary("c++", LibraryKind::Library));
+  else if (Target.isOSLinux())
```
Comment (ian-twilightcoder) [inline lib/AST/ModuleDependencies.cpp:521]: I thought Android used libc++ and not libstdc++?

File: lib/AST/ModuleDependencies.cpp [hunk 5]
```diff
@@ -509,6 +510,86 @@ swift::dependencies::checkImportNotTautological(const ImportPath::Module moduleP
...
+                    })) {
+    // Only link with CxxStdlib on platforms where the overlay is available.
+    switch (Target.getOS()) {
+    case llvm::Triple::Linux:
+      if (!Target.isAndroid())
+        RegistrationCallback(LinkLibrary("swiftCxxStdlib",
+                                         LibraryKind::Library));
+      break;
+    case llvm::Triple::Win32: {
+      bool isStatic = hasStaticCxxStdlib;
```
Comment (ian-twilightcoder) [inline lib/AST/ModuleDependencies.cpp:546]: What is this variable for?

**Review by ian-twilightcoder — APPROVED** [review]

File: lib/AST/ModuleDependencies.cpp (hunk 3 above)
Comment (artemcm) [reply lib/AST/ModuleDependencies.cpp:187]: Removed. Thanks.

File: lib/AST/ConformanceLookupTable.cpp (hunk 2 above)
Comment (artemcm) [reply lib/AST/ConformanceLookupTable.cpp:311]: Not. Thanks for catching.

File: lib/AST/ModuleDependencies.cpp (hunk 5 above)
Comment (artemcm) [reply lib/AST/ModuleDependencies.cpp:546]: Vestige of debugging. Removed, thanks.

File: lib/AST/ModuleDependencies.cpp (hunk 4 above)
Comment (artemcm) [reply lib/AST/ModuleDependencies.cpp:521]: The core logic here just moved from how it was in `GenDecl.cpp`. But @egorzhdan may know the answer.

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1622
https://github.com/apple/llvm-project/pull/8810
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1622
https://github.com/apple/llvm-project/pull/8810
@swift-ci test

File: include/swift/AST/ModuleDependencies.h [hunk 6]
```diff
@@ -93,12 +95,10 @@ struct ModuleDependencyID {
   std::string ModuleName;
   ModuleDependencyKind Kind;
   bool operator==(const ModuleDependencyID &Other) const {
-    return std::tie(ModuleName, Kind) ==
-           std::tie(Other.ModuleName, Other.Kind);
+    return std::tie(ModuleName, Kind) == std::tie(Other.ModuleName, Other.Kind);
```
Comment (cachemeifyoucan) [inline include/swift/AST/ModuleDependencies.h:98]: Nit: avoid re-formatting unchanged logic. This makes review and git blame hard.

**Review by cachemeifyoucan — COMMENTED** [review]
What do you think about the idea that forceLinked library should be pass to linker as full path so linker will not drop it when nothing from it is referenced? (Might need to check linker semantics for all the supported platforms/linkers). Then we can drop FORCE_LINK symbols too when explicit auto link is used.

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1622
@swift-ci test

**Comment by artemcm** [comment]
https://github.com/apple/swift-driver/pull/1622
@swift-ci test

**Pull request merged**